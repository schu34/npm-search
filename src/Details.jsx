import React from "react";
import { useInput, Text, Box } from "ink";

const Details = ({ details }) => {
  return (
    <Box flexDirection={"column"}>
      {details.split("\n").map(line => (
        <Text>{line}</Text>
      ))}
    </Box>
  );
};

export default Details;
