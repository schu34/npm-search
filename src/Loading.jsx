import React from "react";
import PropTypes from "prop-types";
import { Color } from "ink";

const Loading = ({ message }) => <Color blue> {message} </Color>;

Loading.propTypes = {
  message: PropTypes.string.isRequired
};

export default Loading;
