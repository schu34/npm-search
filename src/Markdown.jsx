import React, { useState, useEffect } from "react";
import { Box, useInput, Text } from "ink";

const Markdown = ({ source, width, height }) => {
  const [formattedSource, setFormattedSource] = useState([]);
  const [top, setTop] = useState(1);

  useEffect(() => {
    const splitSource = source.split("\n");
    const lines = splitSource.reduce((result, nextLine, lineNumber) => {
      const linesToAppend = [];
      const words = nextLine.split(" ");
      let currentLine = "";
      for (let i = 0; i < words.length; i++) {
        const word = words[i];

        if (currentLine.length + word.length + 1 <= width) {
          currentLine += ` ${word}`;
        } else {
          /* TODO: currently assuming that the window won't be
           * so narrow that one word is too wide for a line.
           * this should be handled better though (maybe we can hyphenate?)
           */
          linesToAppend.push(
            <Text key={currentLine + i + lineNumber}>{currentLine}</Text>
          );
          currentLine = word;
        }
      }
      if (currentLine.length)
        linesToAppend.push(
          <Text key={currentLine + "last" + lineNumber}>{currentLine}</Text>
        );

      return result.concat(linesToAppend);
    }, []);
    setFormattedSource(lines);
  }, [source]);

  useInput((input, key) => {
    if (key.upArrow || input === "k") {
      return setTop(Math.max(top - 1, 0));
    }

    if (key.downArrow || input === "j") {
      return setTop(Math.min(formattedSource.length - height, top + 1));
    }
  });

  return (
    <Box flexDirection={"column"} width={width} height={height}>
      {formattedSource.slice(top, top + height - 1)}
    </Box>
  );
};

export default Markdown;
