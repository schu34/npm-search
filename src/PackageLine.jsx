import React from "react";
import { Text } from "ink";
import PropTypes from "prop-types";

const PackageLine = ({ packageName, ...props }) => {
  return <Text {...props}>{packageName}</Text>;
};

PackageLine.propTypes = {
  packageName: PropTypes.string.isRequired,
};

export default PackageLine;
