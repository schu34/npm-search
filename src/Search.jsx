import React from "react";
import PropTypes from "prop-types";
import { useInput } from "ink";
import TextInput from "ink-text-input";

const Search = ({ value, onChange, getResults }) => {
  useInput((_, key) => {
    if (key.return) getResults();
  });

  return (
    <TextInput
      placeholder={"search for a package"}
      value={value}
      onChange={onChange}
    />
  );
};

Search.propTypes = {
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  getResults: PropTypes.func.isRequired,
};
export default Search;
