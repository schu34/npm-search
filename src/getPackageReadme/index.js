import fetch from "node-fetch";
import gh from "github-url-to-object";
var GITHUB_README = "https://raw.githubusercontent.com/%s";

export default async function getPackageReadme(packageName) {
  const npmUrl = `https://registry.npmjs.org/${packageName}`;
  const data = await fetch(npmUrl);
  const parsedData = await data.json();
  if (parsedData.readme) {
    return parsedData.readme;
  }
  console.log(parsedData);

  const { readmeFilename } = parsedData;
  if (!readmeFilename) {
    throw new Error(`${packageName}: package.json has no readmeFilename`);
  }

  const repoUrl = parsedData.repository && parsedData.repository.url;
  if (!repoUrl) {
    throw new Error(`${packageName}: package.json has no repository`);
  }
  const repoObj = gh(repoUrl);
  if (!repoObj || !repoObj.user || !repoObj.repo) {
    throw new Error(`${packageName}: cannot parse repository url`);
  }
  const { user, repo } = repoObj;

  const readmePath = `${user}/${repo}/master/${readmeFilename}`;
  const githubUrl = GITHUB_README.replace("%s", readmePath);

  const readmeData = await fetch(githubUrl);
  return readmeData.toString();
}
