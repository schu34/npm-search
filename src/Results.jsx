import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { Box, useInput, Color, Text } from "ink";
import npmSearch from "libnpmsearch";
import { getScreenHeight } from "./utils";
import Loading from "./Loading";

const clamp = (number, min, max) => Math.min(max, Math.max(number, min));

Results.propTypes = {
  search: PropTypes.string.isRequired,
  select: PropTypes.func.isRequired
};

const Results = ({ search, select }) => {
  const [top, setTop] = useState(0);
  const [selection, setSelection] = useState(0);
  const [results, setResults] = useState([]);
  const [loading, setLoading] = useState(true);

  const maxRows = Math.min(getScreenHeight() - 2, results.length);

  useEffect(() => {
    setLoading(true);
    npmSearch(search)
      .then(returnedResults => {
        setLoading(false);
        setResults(returnedResults);
        // setScreen("results");
      })
      .catch(() => {
        setLoading(false);
        setResults([{ name: "error" }]);
        setLoading(false);
      });
  }, [search]);

  useInput((input, key) => {
    if (loading) return;
    if (input === "j" || key.downArrow) {
      if (selection === maxRows - 1 && top + maxRows !== results.length) {
        setTop(top + 1);
      } else {
        setSelection(clamp(selection + 1, 0, maxRows - 1));
      }
    }
    if (input === "k" || key.upArrow) {
      if (selection === 0 && top !== 0) {
        setTop(top - 1);
      } else {
        setSelection(clamp(selection - 1, 0, maxRows - 1));
      }
    }
    if (key.return) {
      select(results[selection].name);
    }
  });

  return loading ? (
    <Loading message="searching for packages..." />
  ) : (
    <Box width={"100%"} height={"100%"} flexDirection="column">
      {results.slice(top, top + maxRows).map((res, i) => {
        return i === selection ? (
          <Box>
            <Color blue>‚ùØ </Color>
            <Text bold>{res.name}</Text>
          </Box>
        ) : (
          "  " + res.name
        );
      })}
      <Box>
        {top + maxRows} {results.length}
      </Box>
    </Box>
  );
};

export default Results;
