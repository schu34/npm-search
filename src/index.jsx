import React, { useState, useEffect } from "react";
import { render, Box, Text, useInput, useApp, Color } from "ink";
import TextInput from "ink-text-input";
import npmSearch from "libnpmsearch";
import { getScreenWidth, getScreenHeight } from "./utils";
import Results from "./Results";

const Search = ({ value, onChange, doSearch }) => {
	useInput((_, key) => {
		if (key.return) doSearch();
	});

	return (
		<TextInput
			placeholder={"search for a package"}
			value={value}
			onChange={onChange}
		/>
	);
};

const Details = ({ text }) => {
	useInput(() => {});
	return <Text>details: {text}</Text>;
};

const App = () => {
	const [search, setSearch] = useState("");
	const [loading, setLoading] = useState(false);
	const [results, setResults] = useState(null);
	const [selection, setSelection] = useState(null);
	const { exit } = useApp();

	const doSearch = () => {
		setLoading(true);
		npmSearch(search)
			.then(returnedResults => {
				setResults(returnedResults);
				setLoading(false);
			})
			.catch(e => {
				setResults([{ name: "error" }]);
				setLoading(false);
			});
	};

	return (
		<Box
			width={getScreenWidth()}
			height={getScreenHeight()}
			flexDirection={"column"}
		>
			{loading ? (
				"loading..."
			) : results ? (
				selection === null ? (
					<Results results={results} select={setSelection} />
				) : (
					<Details text={results[selection].description} />
				)
			) : (
				<Search value={search} onChange={setSearch} doSearch={doSearch} />
			)}
		</Box>
	);
};

render(<App />, { experimental: true });
