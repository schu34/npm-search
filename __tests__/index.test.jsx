import React from "react";
import {App} from "../src";
import { render } from "ink-testing-library";

describe("<App />", () => {
  it("should render without errors", () => {
    const { lastFrame } = render(<App />);
    expect(lastFrame()).toMatchSnapshot();
  });
});