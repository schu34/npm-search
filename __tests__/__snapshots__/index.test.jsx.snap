// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render without errors 1`] = `"[2msearch for a package[22m"`;
